### PR TITLE
[Opt](cloud) Add segment prefetcher

### PR DESCRIPTION
**Pull request by bobhan1**
### Cold run test result
- Environment: 1 FE + 2 BE(aliyun, [ecs.i4g.16xlarge](https://help.aliyun.com/zh/ecs/user-guide/instance-families-with-local-ssds#i4g), 2.7 GHz Intel® Xeon® Scalable (Ice Lake) processors with high-performance local NVMe SSDs)
    - Network baseline bandwidth **32 Gbit/s** on each machine
- Duplicate table, 16 buckets, 6178831001 rows, data size is 396.837 GB
- Cold run means **query without any doris page, OS page cache or local disk file cache, all data will be fetched from S3**.
- Query: `select count(user_id) from test_dup3 where env_id % 10 = 0 and user_id%10=0 and time % 10 =0 and event_id % 10 = 0 and length(domain) % 10 = 0 and length(path) % 10 = 0 and length(query)%10=0;`
    - For this query, each BE will download **114GB** data to local disk file cache from S3
    - 8 instance * 48 scanner = 384 scanner on each BE, and scanner thread pool size is set to 48

||no prefetch|prefetch_block_size=2|
|---|---|---|
| query time | 2min47sec | 39sec273ms |
| network bandwidth</br>(on one BE) | slow| fast|

based on https://github.com/apache/doris/pull/59476
### Check List (For Author)

- Test 
    - [ ] Regression test
    - [ ] Unit Test
    - [ ] Manual test (add detailed scripts or steps below)
    - [ ] No need to test or manual test. Explain why:
        - [ ] This is a refactor/code format and no logic has been changed.
        - [ ] Previous test can cover this change.
        - [ ] No code files have been changed.
        - [ ] Other reason 

- Behavior changed:
    - [ ] No.
    - [ ] Yes. 

- Does this need documentation?
    - [ ] No.
    - [ ] Yes. 

### Check List (For Reviewer who merge this PR)

- [ ] Confirm the release note
- [ ] Confirm test cases
- [ ] Confirm document
- [ ] Add branch pick label

### CODE REVIEW

**Comment by Thearas** [comment]
Thank you for your contribution to Apache Doris.
Don't know what should be done next? See [How to process your PR](https://cwiki.apache.org/confluence/display/DORIS/How+to+process+your+PR).

Please clearly describe your PR:
1. What problem was fixed (it's best to include specific error reporting information). How it was fixed.
2. Which behaviors were modified. What was the previous behavior, what is it now, why was it modified, and what possible impacts might there be.
3. What features were added. Why was this function added?
4. Which code was refactored and why was this part of the code refactored?
5. Which functions were optimized and what is the difference before and after the optimization?

**Comment by hello-stephen** [comment]
# BE Regression && UT Coverage Report
Increment line coverage `26.61% (161/605)` :tada:

[Increment coverage report](http://coverage.selectdb-in.cc/coverage/59482_f2efab26178bd89c271613b1c7e917122c13eadd_merge/increment_report/index.html)
[Complete coverage report](http://coverage.selectdb-in.cc/coverage/59482_f2efab26178bd89c271613b1c7e917122c13eadd_merge/report/index.html)
| Category          | Coverage           |
|-------------------|--------------------|
| Function Coverage | 72.11% (25376/35193) |
| Line Coverage     | 58.85% (265471/451064) |
| Region Coverage   | 53.48% (219697/410840) |
| Branch Coverage   | 55.21% (94502/171158) |

**Comment by dataroaring** [comment]
# Code Review: [Opt](cloud) Add segment prefetcher

## Summary

This PR introduces a **segment prefetcher** to optimize cold query performance when reading data from S3 in cloud deployments. The benchmarks are impressive: **~4.3x improvement** (2min 47sec → 39sec) for cold queries fetching 114GB per BE from S3.

The prefetcher uses a **read-ahead windowing strategy** - it maintains awareness of upcoming blocks based on the row bitmap and triggers async prefetch when the read position approaches block boundaries. This is particularly effective for columnar storage where predicate filtering creates non-contiguous row access patterns.

---

## Architecture Assessment

### Design Strengths

1. **Lazy Initialization**: Prefetching is only enabled for cloud mode and specific reader types (query/compaction), minimizing overhead for local deployments.

2. **Weak Reference Pattern** (`cached_remote_file_reader.cpp:627-642`): The prefetch task captures a `weak_ptr` to avoid preventing destruction of the file reader - good design.

3. **Complex Type Handling**: Properly handles Map/Array/Struct columns where nested column access patterns differ from outer offset columns.

---

## Critical Issues

### 1. Race Condition in Block Sequence Access (HIGH)

In `segment_prefetcher.cpp`, `need_prefetch()` modifies `_prefetched_idx` and `_current_block_idx` without synchronization, while these could potentially be accessed from prefetch callbacks during verbose logging:

```cpp
bool SegmentPrefetcher::need_prefetch(uint32_t rowid, std::vector<BlockRange>* ranges) {
    // ... modifies _prefetched_idx, _current_block_idx without lock
}
```

**Recommendation**: Either add mutex protection around shared state, or document that `need_prefetch()` must only be called from the scan thread (and ensure prefetch callbacks never access this state).

### 2. Thread Pool Starvation Risk (MEDIUM)

Configuration allows up to **2000 threads** in the prefetch pool. With high concurrency and large `prefetch_block_size`, this could exhaust system resources.

```cpp
DEFINE_Int64(segment_prefetch_thread_pool_thread_num_max, "2000");
```

**Recommendation**: Consider adding a bounded queue with back-pressure, or adaptive prefetch rate based on cache hit ratio.

### 3. Missing Error Propagation (MEDIUM)

Prefetch failures are silently logged at VLOG_DEBUG level but never surfaced:

```cpp
if (!st.ok()) {
    VLOG_DEBUG << "Failed to submit prefetch task...";
}
```

**Recommendation**: Add metrics/counters for prefetch failures to enable monitoring.

---

## Code Quality Issues

### 4. Inconsistent Prefetch Window Calculation (`segment_iterator.cpp:573`)

```cpp
int window_size = 1 + (is_query ? config::query_segment_file_cache_prefetch_block_size
                                : config::compaction_segment_file_cache_prefetch_block_size);
```

The `1 +` is confusing. If `prefetch_block_size=2`, this creates a window of 3. The naming suggests "how many blocks to prefetch ahead" but the +1 makes it "current + N ahead".

**Recommendation**: Either rename the config to `prefetch_window_size` or document the +1 semantics.

### 5. Potential Dangling IOContext (`cached_remote_file_reader.cpp:618-624`)

The copied `dryrun_ctx` nullifies `query_id`, `file_cache_stats`, `file_reader_stats` but copies other fields that might contain dangling references after the parent context is destroyed:

```cpp
IOContext dryrun_ctx;
if (io_ctx != nullptr) {
    dryrun_ctx = *io_ctx;  // shallow copy
}
dryrun_ctx.query_id = nullptr;  // only nullifies some fields
```

**Recommendation**: Review `IOContext` for other pointer/reference members that need clearing.

---

## Testing Gaps

### 6. No Unit Tests

The PR adds ~1150 lines of new code but no regression or unit tests are included. Given the complexity of block sequence building, prefetch window management, and complex type handling, unit tests would be valuable for:
- `SegmentPrefetcher::add_rowids()` with various bitmap patterns
- `need_prefetch()` boundary conditions
- Backward iteration logic

### 7. No Compaction Testing Validation

Compaction prefetch is enabled but the benchmark only covers query workloads. Compaction has different I/O patterns (sequential, full segment reads).

---

## Minor Issues

1. **Magic Numbers**: `segment_file_cache_consume_rowids_batch_size = 8000` lacks rationale documentation.

2. **Verbose Logging Pattern**: `LOG_IF(INFO, config::enable_segment_prefetch_verbose_log)` is repeated ~20 times - consider a macro.

---

## Configuration Review

| Config | Default | Assessment |
|--------|---------|------------|
| `enable_query_segment_file_cache_prefetch` | `false` | Good - conservative default |
| `query_segment_file_cache_prefetch_block_size` | `2` | Reasonable |
| `segment_prefetch_thread_pool_thread_num_max` | `2000` | Potentially too high |

---

## Verdict

**Overall**: Well-structured optimization with significant performance benefits for cloud deployments. The core design is sound.

**Recommendations before merge**:
1. Address or document the thread safety model
2. Add basic unit tests for the prefetcher
3. Add metrics for prefetch success/failure rates
4. Consider lowering `segment_prefetch_thread_pool_thread_num_max` default or adding back-pressure

File: be/src/util/concurrency_stats.h [hunk 1]
```diff
@@ -0,0 +1,128 @@
...
+    ConcurrencyStatsManager(const ConcurrencyStatsManager&) = delete;
+    ConcurrencyStatsManager& operator=(const ConcurrencyStatsManager&) = delete;
+
+    void _dump_thread_func();
+
+    // All counters in the order they should be printed
+    std::vector<ConcurrencyCounter*> _counters;
+
+    std::atomic<bool> _running;
+    std::unique_ptr<std::thread> _dump_thread;
```
Comment (gavinchou) [inline be/src/util/concurrency_stats.h:115]: better name the thread with something like "set_thread_name_p()"

File: be/src/util/concurrency_stats.cpp [hunk 2]
```diff
@@ -0,0 +1,131 @@
...
+#include <chrono>
+#include <sstream>
+
+#include "common/config.h"
+#include "common/logging.h"
+
+namespace doris {
+ConcurrencyStatsManager::ConcurrencyStatsManager() : _running(false) {
+    // Initialize all counters in the order of read path (top to bottom)
+    vscanner_get_block = new ConcurrencyCounter("vscanner");
```
Comment (gavinchou) [inline be/src/util/concurrency_stats.cpp:30]: when will these counters be reclaimed?

**Review by gavinchou — DISMISSED** [review]

File: be/src/util/concurrency_stats.cpp (hunk 2 above)
Comment (bobhan1) [reply be/src/util/concurrency_stats.cpp:30]: in `~ConcurrencyStatsManager()`

**Review by gavinchou — DISMISSED** [review]

**Comment by dataroaring** [comment]
### Code review

No issues found. Checked for bugs and CLAUDE.md compliance.

🤖 Generated with [Claude Code](https://claude.ai/code)

<sub>- If this code review was useful, please react with 👍. Otherwise, react with 👎.</sub>

**Review by gavinchou — APPROVED** [review]

**Review by liaoxin01 — APPROVED** [review]
LGTM

**Review by dataroaring — APPROVED** [review]
LGTM
